module/module.go

```diff
@@ -0,0 +1,422 @@
+			Subtype: protoutils.ResourceNameToProto(resource.Name{
+				Subtype: s.Subtype,
+				Name:    "",
+			}),
+			ProtoService: s.SvcName,
+		}
+
+		handler := &pb.HandlerDefinition{Subtype: subtype}
+		for _, m := range models {
+			handler.Models = append(handler.Models, m.String())
+		}
+		pMap.Handlers = append(pMap.Handlers, handler)
+	}
+	return pMap
+}
+
+// NewHandlerMapFromProto converts protobuf to HandlerMap.
+func NewHandlerMapFromProto(ctx context.Context, pMap *pb.HandlerMap, conn *grpc.ClientConn) (HandlerMap, error) {
+	hMap := make(HandlerMap)
+	refClient := grpcreflect.NewClient(ctx, reflectpb.NewServerReflectionClient(conn))
+	defer refClient.Reset()
+	reflSource := grpcurl.DescriptorSourceFromServer(ctx, refClient)
+
+	var errs error
+	for _, h := range pMap.GetHandlers() {
+		api := protoutils.ResourceNameFromProto(h.Subtype.Subtype).Subtype
+
+		symDesc, err := reflSource.FindSymbol(h.Subtype.ProtoService)
+		if err != nil {
+			errs = multierr.Combine(errs, err)
+			continue
+		}
+		svcDesc, ok := symDesc.(*desc.ServiceDescriptor)
+		if !ok {
+			return nil, errors.Errorf("expected descriptor to be service descriptor but got %T", symDesc)
+		}
+		subtype := &resource.RPCSubtype{
+			Subtype: api,
+			Desc:    svcDesc,
+		}
+		for _, m := range h.Models {
+			model, err := resource.NewModelFromString(m)
+			if err != nil {
+				return nil, err
+			}
+			hMap[*subtype] = append(hMap[*subtype], model)
+		}
+	}
+	return hMap, errs
+}
+
+// Module represents an external resource module that services components/services.
+type Module struct {
+	parent                  *client.RobotClient
+	grpcServer              *grpc.Server
+	logger                  *zap.SugaredLogger
+	mu                      sync.Mutex
+	ready                   bool
+	addr                    string
+	parentAddr              string
+	activeBackgroundWorkers sync.WaitGroup
+	handlers                HandlerMap
+	services                map[resource.Subtype]subtype.Service
+	pb.UnimplementedModuleServiceServer
+}
+
+// NewModule returns the basic module framework/structure.
+func NewModule(address string, logger *zap.SugaredLogger) *Module {
+	m := &Module{
+		logger:     logger,
+		addr:       address,
+		grpcServer: grpc.NewServer(),
+		ready:      true,
+		handlers:   make(HandlerMap),
+		services:   make(map[resource.Subtype]subtype.Service),
+	}
+	m.buildHandlerMap()
+	pb.RegisterModuleServiceServer(m.grpcServer, m)
+	reflection.Register(m.grpcServer)
+	return m
+}
+
+// Start starts the module service and grpc server.
+func (m *Module) Start(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	err := m.initSubtypeServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	oldMask := syscall.Umask(0o077)
+	lis, err := net.Listen("unix", m.addr)
+	syscall.Umask(oldMask)
+	if err != nil {
+		return errors.WithMessage(err, "failed to listen")
+	}
+
+	m.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
```

Comment:
should this be a ManagedGo?

Reply:
I don't believe so. If I understand correctly what MangedGo does, it would basically try to restart it if it panicked, and I don't think that's a good thing in this case, especially as this is only the grpc listner function. There's no promises that other states are correct, so it could quickly block on an unreleased mutex or something in the module's custom areas. I think the proper handling would likely be restarting the module itself (the parent would do this automatically) and as part of the future PR where reconfiguring an existing module is handled, reloading the config into a restarted module should be basically the same thing.